src/input.rs

```diff
@@ -242,6 +242,30 @@ where
                 }
             });
     }
+
+    fn update(
+        &self,
+        db: &mut <Q as QueryDb<'_>>::DynDb,
+        key: &<Q as Query>::Key,
+        value_fn: &mut dyn FnMut(&mut Q::Value),
+        durability: Durability,
+    ) {
+        db.salsa_runtime_mut()
+            .with_incremented_revision(&mut |next_revision| {
+                let slots = self.slots.write();
+
+                let slot = slots
+                    .get(key)
+                    .unwrap_or_else(|| panic!("no value set for {:?}({:?})", Q::default(), key));
+
+                let mut slot_stamped_value = slot.stamped_value.write();
+                let old_durability = slot_stamped_value.durability;
+                value_fn(&mut slot_stamped_value.value);
```

Comment:
What happens in the event of a panic? That seems like a problem (and the question should be added to the RFC).

Reply:
I guess the answer is probably that the panic propagates out and the input is in an invalid state. In other words, unlike for derived queries, we can't recover from panics in the mutator thread. This seems reasonable, but we should document it. (At least, I don't see how we could recover.)

Reply:
Agreed, I'll add to the RFC. For consistency with the other RFC we would unmemoize the value, by calling `take` then `value_fn`. This works on the other RFC because it just means the derived query re-calculates the value. But a missing input query is different from an unmemoized derived value, as a missing input query will panic the next time.